skip.js

```diff
@@ -1,18 +1,70 @@
+// Create an observer instance linked to the callback function
+let observer = null;
+
 function skip(button, inner_text) {
     let buttons = document.querySelectorAll(button);
 
     // Confirm we aren't about to click on something we don't intend to.
     for(let i = 0;i < buttons.length; ++i) {
-        if(buttons[i].innerText === inner_text)
+        if (buttons[i].innerText === inner_text) {
+            if (observer != null) {
+                observer.disconnect(); // with mission accomplished, deactivate observer
+                observer = null;
+            }
             buttons[i].click();
+        }
     }
 }
 
-switch (window.location.hostname) {
+function go() {
+    switch (window.location.hostname) {
     case 'delivery.cardago.com':
         skip('.btn_base', 'READY TO OPEN YOUR GIFT CARD?');
         break;
     case 'www.paypal.com':
         skip('[data-pa-click]', 'Skip');
         break;
+    }
+}
+
+function configure_dom_observer(id_to_observe) {
+    // Select the node that will be observed for mutations
+    let targetNode = document.getElementById(id_to_observe);
+
+    // Options for the observer (which mutations to observe)
+    const config = { attributes: true, childList: true, subtree: true };
+
+    // Callback function to execute when mutations are observed
+    const callback = function(mutationsList, observer) {
+        // Use traditional 'for loops' for IE 11
+        for(let mutation of mutationsList) {
+            if (mutation.type === 'childList') {
+                //console.log('A child node has been added or removed.');
+                go();
+            }
+            else if (mutation.type === 'attributes') {
+                //console.log('The ' + mutation.attributeName + ' attribute was modified.');
+                // don't care about attribute mutations for now
+            }
+        }
+    };
+
+    observer = new MutationObserver(callback);
+    // Start observing the target node for configured mutations
+    observer.observe(targetNode, config);
+}
+
+function init() {
+    switch (window.location.hostname) {
+        case 'delivery.cardago.com':
+            configure_dom_observer('content-main');
```

Comment:
Afaict, `content-main` doesn't mutate at all on MGCP GC. I couldn't get a mutation observer to fire when I added it by hand in the console, do you know if it normally would if it was part of the extension? Otherwise this might never click the button.

Reply:
So yes you're right, attaching an observer for MGCP is overkill since it doesn't mutate like PPDG. But I like creating one-size-fits-all since the logic just gets simpler, especially if more sites need to be supported in the future. The flow of the logic is in init, we explicitly call go() at the end after the observer has been instantiated anyways, so there's always a first attempt to click a matching button if one exists. For PPDG, this go() call will do nothing since the buttons array will be empty when we explicitly call it, and then later go() will be called for each mutation, which eventually will have the matching button in the DOM (and it may also be called a few more times if there are mutations prior to the Skip button appearing). 

On the flip side, MGCP we set up an observer which will immediately be disconnected since the matching button will be found in the first attempt. IMO the computational overhead of calling go() superfluously and/or setting up an observer that will never observe a mutation is a worthy tradeoff for simpler code logic.

Reply:
Agreed, I had missed the go() call at the end of this function when I had commented. I definitely strive to avoid having any site specific things as well, making things as general as possible is the goal, so I like what you were going for here.

The one thing that I'm not sure of is how to avoid the observer being global in that case, since we wouldn't have the callback getting the observer to pass to go. Although thinking about it now, seems like we could probably just create the observer in init() and pass it to configure_dom_observer() and go() as necessary and it wouldn't be a problem.